### ISSUE TITLE
Build fails with Error FS2014 when a signature file is used with IWSAMs

### ISSUE THREAD

**Issue by jrharnden** [dotnet/fsharp#16329]
This appears to be directly related to #15919. The fix identified by @vzarytovskii and implemented by @gusty works for the scenario 
below when there is no signature file. When a signature file is added it still triggers the error. The types do not even need to be in the signature file to trigger the error.

Error:

> A problem occurred writing the binary 'obj\Debug\net7.0\refint\Foo.dll': Error in pass3 for type Foo, error: Error in pass3 for type CompilerGoesBoom`1, error: Error in GetMethodRefAsMethodDefIdx for mref = ("Foo.IHasStaticAbstractBase<'a>.Boom", "CompilerGoesBoom`1"), error: MethodDefNotFound

Scenario identified by @En3Tho:

```
type IHasStaticAbstractBase<'a> =
    static abstract Boom: unit -> 'a

type IHasStaticAbstractChild<'a> =
    inherit IHasStaticAbstractBase<'a>

type CompilerGoesBoom<'a>() =
    interface IHasStaticAbstractChild<'a> with // this goes boom because inheritance is involved
        static member Boom() = Unchecked.defaultof<'a>

type CompilerDoesNotGoBoom<'a>() =
    interface IHasStaticAbstractBase<'a> with // this does not go boom, direct base interface
        static member Boom() = Unchecked.defaultof<'a>
```

**Comment by vzarytovskii** [comment]
Hm. Interesting. @jrharnden is this on the main?

**Comment by jrharnden** [comment]
I backported the changes to fsharp-release-dev17.7 prior to the .net 8 release. I can check on main real quick.

**Comment by vzarytovskii** [comment]
I have no idea currently why it's failing, but might be some weird case of us merging empty signature and implementation and not doing fixup or something.

**Comment by jrharnden** [comment]
It looks like the same issue occurs on the main branch. To clarify, triggering the bug doesn't require the signature to be empty, just for a signature file to exist. It still fails with the signature below.

```
type IHasStaticAbstractBase<'a> =
    static abstract Boom: unit -> 'a

[<Interface>]
type IHasStaticAbstractChild<'a> =
    inherit IHasStaticAbstractBase<'a>

[<Class>]
type CompilerGoesBoom<'a> =
    interface IHasStaticAbstractChild<'a>

[<Class>]
type CompilerDoesNotGoBoom<'a> =
    interface IHasStaticAbstractBase<'a>
```

**Comment by vzarytovskii** [comment]
> It looks like the same issue occurs on the main branch. To clarify, triggering the bug doesn't require the signature to be empty, just for a signature file to exist. It still fails with the signature below.
> 
> 
> 
> ```
> 
> type IHasStaticAbstractBase<'a> =
> 
>     static abstract Boom: unit -> 'a
> 
> 
> 
> [<Interface>]
> 
> type IHasStaticAbstractChild<'a> =
> 
>     inherit IHasStaticAbstractBase<'a>
> 
> 
> 
> [<Class>]
> 
> type CompilerGoesBoom<'a> =
> 
>     interface IHasStaticAbstractChild<'a>
> 
> 
> 
> [<Class>]
> 
> type CompilerDoesNotGoBoom<'a> =
> 
>     interface IHasStaticAbstractBase<'a>
> 
> ```

Yeah, I didn't mean to emphasize on empty signature, just signature. We probably do something different when we merge those.

Will need to spend some time to figure it out.

**Comment by vzarytovskii** [comment]
@jrharnden I can't seem to reproduce it:
```shell
> cat repro.fs
```
```fsharp
namespace Foobar
type IHasStaticAbstractBase<'a> =
    static abstract Boom: unit -> 'a

type IHasStaticAbstractChild<'a> =
    inherit IHasStaticAbstractBase<'a>

type CompilerGoesBoom<'a>() =
    interface IHasStaticAbstractChild<'a> with
        static member Boom() = Unchecked.defaultof<'a>
```
```shell
> cat repro.fsi
```
```fsharp
namespace Foobar
```
```shell                                                                                                                                                                                                                                                                                                                                               
> cat repro.rsp
```
```shell
--target:library
--targetprofile:netstandard
--simpleresolution
repro.fsi
repro.fs
```
```shell                                                                                                                                                                                                                                                                                                                                                       
> dotnet artifacts/bin/fsc/Debug/net8.0/fsc.dll @repro.rsp
Microsoft (R) F# Compiler version 12.8.0.0 for F# 8.0
Copyright (c) Microsoft Corporation. All Rights Reserved.

/Users/u/code/fsharp/repro.fs(3,5): warning FS3535: Declaring "interfaces with static abstract methods" is an advanced feature. See https://aka.ms/fsharp-iwsams for guidance. You can disable this warning by using '#nowarn "3535"' or '--nowarn:3535'.
```
```shell
> file repro.dll
repro.dll: PE32 executable (DLL) (console) Intel 80386 Mono/.Net assembly, for MS Windows
```

**Comment by vzarytovskii** [comment]
Same with the following signature file:
```fsharp
type IHasStaticAbstractBase<'a> =
    static abstract Boom: unit -> 'a

[<Interface>]
type IHasStaticAbstractChild<'a> =
    inherit IHasStaticAbstractBase<'a>

[<Class>]
type CompilerGoesBoom<'a> =
    interface IHasStaticAbstractChild<'a>

```

**Comment by vzarytovskii** [comment]
```shell
> git rev-parse HEAD
```
[2a25184293e39a635217670652b00680de04472a](https://github.com/dotnet/fsharp/commit/2a25184293e39a635217670652b00680de04472a)

**Comment by jrharnden** [comment]
Sorry. I've been running it directly from visual studio. I've attached a minimal repro setup for this method. I need to go do family stuff for now, but I can try to get a command line repro later tonight. Excluding the signature file from the solution should make the bug go away. Compiler is copied directly from artifacts\bin\fsc\Release\net8.0 on the same commit you as you have above.

**Comment by vzarytovskii** [comment]
> 
> 
> [Foobar.zip](https://github.com/dotnet/fsharp/files/13452938/Foobar.zip)
> 
> Sorry. I've been running it directly from visual studio. I've attached a minimal repro setup for this method. I need to go do family stuff for now, but I can try to get a command line repro later tonight. Excluding the signature file from the solution should make the bug go away. Compiler is copied directly from artifacts\bin\fsc\Release\net8.0 on the same commit you as you have above.
> 
> 

In vs meaning, you see it in ide, or build it in ide and see output? Also, where do you replace the compiler?

**Comment by jrharnden** [comment]
I open that attached project in the latest Visual Studio 2022 community edition and build it using f5. The copy of the compiler mentioned is in the top-level tools directory. It is chosen using the property group below in the project file. The error should show up in the error window within visual studio once it builds. The build output text leads me to believe that the correct compiler is being used and should show the compiler options chosen as well.

  <PropertyGroup>
    <FSharp_Shim_Present>false</FSharp_Shim_Present>
    <FscToolPath>$(SolutionDir)tools</FscToolPath>
    <FscToolExe>fsc.exe</FscToolExe>
    <DotnetFscCompilerPath></DotnetFscCompilerPath>
    <FSharpPreferNetFrameworkTools>False</FSharpPreferNetFrameworkTools>
    <FSharpPrefer64BitTools>True</FSharpPrefer64BitTools>
  </PropertyGroup>

**Comment by jrharnden** [comment]
It looks like the problem is related to reference assemblies which I guess are enabled by default in Visual Studio. I can reproduce it using the same commands you were using and the rsp file below. I can also successfully build the solution in visual studio now if I set the ProduceReferenceAssembly property to false in the project config file. It looks like you are aware of the issue already https://github.com/dotnet/fsharp/issues/15919#issuecomment-1703885152 ? If so, sorry for the wild goose chase. I really appreciate your help though!

```
--target:library
--targetprofile:netstandard
--simpleresolution
--refout:RefRepro.dll
repro.fsi
repro.fs
```

**Comment by vzarytovskii** [comment]
Hm, weird, seems to be different paths for reassemblies and no refassemblies.

So it's probably different issue (I.e. not naming issue). 

@T-Gro could you please point me to the changes you've made to figure out real visibility to generate refassemblies? There's probably a search for a method by the name with no fixups.

**Comment by T-Gro** [comment]
@vzarytovskii : IL real visiblity did not udergo any changes (cannot be done, since there could be a C# project depending on it).
For the F# signature, it is all driven via the custom "SignatureHash" module and logic in it.

Seeing the error, we had cases in ilwrite when reference assembly logic depended on code not being flushed out yet - there is a shared mutable cenv and things have to be done in a sensitive order. Have a look at e.g. GenTypeDefPass2  in ilwrite.fs

**Comment by T-Gro** [comment]
@vzarytovskii : Are you looking at it already, or can I take over later?

**Comment by vzarytovskii** [comment]
> @vzarytovskii : Are you looking at it already, or can I take over later?

Already looking at it. It seems unrelated, it's some fixups we don't do. I'll fix it.

**Comment by vzarytovskii** [comment]
Even more minimal repro:
```fsharp
namespace Foobar
type IHasStaticAbstractBase<'a> =
    static abstract Boom: unit -> 'a

type CompilerGoesBoom<'a>() =
    interface IHasStaticAbstractBase<'a> with
        static member Boom() = Unchecked.defaultof<'a>
```

With empty signature file will result in the error.

**Comment by vzarytovskii** [comment]
Or even simpler:
```fsharp
namespace Foobar
open System.Numerics

type CompilerGoesBoom<'a, 'b when 'a :> IUnaryPlusOperators<'a, 'b>>() =
    interface IUnaryPlusOperators<'a, 'b> with
        static member op_UnaryPlus(a: 'a) : 'b = Unchecked.defaultof<'b>
```
